### PR TITLE
feat: make handshake cancelable (backport #857)

### PR DESCRIPTION
**Pull request by yihuang**
it'll make the handshake work with graceful shutdown(see: https://github.com/cosmos/cosmos-sdk/issues/16202)

handshake could be a long running process if there are many local blocks to replay, for example we use it to do profiling.

---

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: consensus/replay.go [hunk 1]
```diff
@@ -265,7 +271,7 @@ func (h *Handshaker) Handshake(proxyApp proxy.AppConns) error {
 	}
 
 	// Replay blocks up to the latest in the blockstore.
-	_, err = h.ReplayBlocks(h.initialState, appHash, blockHeight, proxyApp)
+	appHash, err = h.ReplayBlocksWithContext(ctx, h.initialState, appHash, blockHeight, proxyApp)
```
Comment (sergio-mena) [inline consensus/replay.go:274]: I know in future branches the original here re-writes `appHash` but not here. I'd leave it as is unless we understand why we are making that change.

Comment (sergio-mena) [reply consensus/replay.go:274]: @yihuang any thoughts?

Comment (yihuang) [reply consensus/replay.go:274]: I think it's kind of bug fix, previously the log don't output latest hash, but I'm ok that we change that in a separate PR

Comment (yihuang) [reply consensus/replay.go:274]: hi @sergio-mena , do you want me to revert this line?

Comment (sergio-mena) [reply consensus/replay.go:274]: Hi @yihuang sorry for delay on coming back to this (constant flow of kinda-more-urgent things)

I took a look at this again, and when comparing this with the backport on v0.34.x (#1012) I realise this change is present there as well, so I'm OK with it. Approving...

**Review by sergio-mena — APPROVED** [review]

**Pull request merged**